src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs

```diff
@@ -0,0 +1,204 @@
+        public BlobBuilder Builder { get; }
+
+        public MethodBodyStreamEncoder(BlobBuilder builder)
+        {
+            if (builder == null)
+            {
+                Throw.BuilderArgumentNull();
+            }
+
+            // Fat methods are 4-byte aligned. We calculate the alignment relative to the start of the ILStream.
+            //
+            // See ECMA-335 paragraph 25.4.5, Method data section:
+            // "At the next 4-byte boundary following the method body can be extra method data sections."
+            if ((builder.Count % 4) != 0)
+            {
+                throw new ArgumentException(SR.BuilderMustAligned, nameof(builder));
+            }
+
+            Builder = builder;
+        }
+
+        /// <summary>
+        /// Encodes a method body and adds it to the method body stream.
+        /// </summary>
+        /// <param name="codeSize">Number of bytes to be reserved for instructions.</param>
+        /// <param name="maxStack">Max stack.</param>
+        /// <param name="exceptionRegionCount">Number of exception regions.</param>
+        /// <param name="hasFatExceptionRegions">True if the exception regions should be encoded in 'fat' format.</param>
+        /// <param name="localVariablesSignature">Local variables signature handle.</param>
+        /// <param name="attributes">Attributes.</param>
+        /// <returns>The offset of the encoded body within the method body stream.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="codeSize"/>, <paramref name="exceptionRegionCount"/>, or <paramref name="maxStack"/> is out of allowed range.
+        /// </exception>
+        public MethodBody AddMethodBody(
+            int codeSize,
+            int maxStack = 8,
+            int exceptionRegionCount = 0,
+            bool hasFatExceptionRegions = false,
+            StandaloneSignatureHandle localVariablesSignature = default(StandaloneSignatureHandle),
+            MethodBodyAttributes attributes = MethodBodyAttributes.InitLocals)
+        {
+            if (codeSize < 0)
+            {
+                Throw.ArgumentOutOfRange(nameof(codeSize));
+            }
+
+            if (unchecked((uint)maxStack) > ushort.MaxValue)
+            {
+                Throw.ArgumentOutOfRange(nameof(maxStack));
+            }
+
+            if (!ExceptionRegionEncoder.IsExceptionRegionCountInBounds(exceptionRegionCount))
+            {
+                Throw.ArgumentOutOfRange(nameof(exceptionRegionCount));
+            }
+
+            int bodyOffset = SerializeHeader(codeSize, (ushort)maxStack, exceptionRegionCount, attributes, localVariablesSignature);
+            var instructions = Builder.ReserveBytes(codeSize);
+
+            var regionEncoder = (exceptionRegionCount > 0) ? 
+                ExceptionRegionEncoder.SerializeTableHeader(Builder, exceptionRegionCount, hasFatExceptionRegions) : default(ExceptionRegionEncoder);
+
+            // return a tuple
```

Comment:
Is this a TODO, then? I thought we settled on proper structs apart from the special case where the order is relevant right? If this is the proposed final shape, then we don't need this comment.